d2/src/test/java/com/linkedin/d2/discovery/util/TestD2Config.java

```diff
@@ -392,6 +392,106 @@ public int getMaxPartitionId()
     assertEquals(0, accessor.getPartitionId(URI.create(illegalUri)));
   }
 
+  // Test PartitionAccessorFactory: match ClassList with custom matching logic
+  @Test
+  public static void testPartitionAccessorFactoryWithCustomMatchingLogic() throws IOException, InterruptedException, URISyntaxException, Exception
+  {
+    @SuppressWarnings("serial")
+    final Map<String,List<String>> clustersData = new HashMap<String,List<String>>()
+    {{
+      put("partitioned-cluster", Arrays.asList(new String[]{"partitioned-service-1", "partitioned-service-2"}));
+    }};
+
+    final Map<String, Object> partitionProperties = new HashMap<>();
+    Map<String, Object> customized = new HashMap<>();
+    List<String> classList = Arrays.asList("TestPartitionAccessor1!Settings", "TestPartitionAccessor2");
+    customized.put("partitionType", "CUSTOM");
+    customized.put("partitionCount", "10");
+    customized.put("partitionAccessorList", classList);
+    partitionProperties.put("partitionProperties", customized);
+
+    final PartitionAccessorRegistry registry = new PartitionAccessorRegistry()
+    {
+      final private Map<String, List<BasePartitionAccessor>> _registry = new HashMap<>();
+
+      @Override
+      public void register(String clusterName, BasePartitionAccessor accessor)
+      {
+        List<BasePartitionAccessor> accessors = _registry.computeIfAbsent(clusterName, k -> new ArrayList<>());
+        accessors.add(accessor);
+      }
+
+      @Override
+      public List<BasePartitionAccessor> getPartitionAccessors(String clusterName)
+      {
+        return _registry.get(clusterName);
+      }
+    };
+
+    class TestPartitionAccessor1 implements PartitionAccessor
+    {
+      @Override
+      public int getPartitionId(URI uri) throws PartitionAccessException
+      {
+        return testGetPartitionId(uri);
+      }
+
+      @Override
+      public int getMaxPartitionId()
+      {
+        return 10;
+      }
+
+      @Override
+      public boolean checkSupportable(String settings) {
+        return "TestPartitionAccessor1!Settings".equals(settings);
```

Comment:
IIUC, you want to use the custom partition accessor if the partition list string **starts with** the class name, instead of equals (the "!Settings" part can be any configs, right?)

Reply:
yes, will update the test here